File: day04/main.rs
fn incrementing_digits(guess: i32) -> bool {
    let mut num = guess;
    let mut curr = 10;

    while num != 0 {
        let next = num % 10;

        if next > curr {
            return false
        }

        curr = next;
        num = num / 10;
    }

    true
}

fn adjacent_digits(guess: i32) -> bool {
    let mut num = guess;
    let mut curr = 10;

    while num != 0 {
        let next = num % 10;

        if next == curr {
            return true
        }

        curr = next;
        num = num / 10;
    }

    false
}

fn digit_pair(guess: i32) -> bool {
    let mut num = guess;
    let mut curr = 10;
    let mut count = 1;

    while num != 0 {
        let next = num % 10;

        if next == curr {
            count += 1;
        } else {
            if count == 2 {
                return true
            }

            count = 1
        }

        curr = next;
        num = num / 10;
    }

    count == 2
}

fn main() {
    let min = 278384;
    let max = 824795;

    let part1 = |w: i32| incrementing_digits(w) && adjacent_digits(w);
    let part2 = |w: i32| incrementing_digits(w) && digit_pair(w);

    let answer1 = (min..max).filter(|&w| part1(w)).count();
    let answer2 = (min..max).filter(|&w| part2(w)).count();

    println!("{}", answer1);
    println!("{}", answer2);
}

File: day04/main.c
#include <stdio.h>
#include <stdlib.h>
#include "../utils.h"

int adjacent_digits(int num) {
  int n = num % 10;

  while((num = num / 10) > 0) {
    if (n == num % 10) return 1;
    n = num % 10;
  }

  return 0;
}

int adjacent_digits_only(int num) {
  int n = num % 10;
  int oc[10] = {0};

  while((num = num / 10) > 0) {
    if (n == num % 10) {
      oc[n]++;
    }
    n = num % 10;
  }

  for(int i = 0; i < 10; i++) {
    if(oc[i] == 1) return 1;
  }

  return 0;
}

int never_decrease(int num) {
  int n = num % 10;

  while((num = num / 10) > 0) {
    if (n < num % 10) return 0;
    n = num % 10;
  }

  return 1;
}

int main(void) {
  // puzzle input
  int min = 278384;
  int max = 824795;

  int a1 = 0, a2 = 0;

  for(int i = min; i < max; i++) {
    if(never_decrease(i))  {
      if(adjacent_digits(i)) a1++;
      if(adjacent_digits_only(i)) a2++;
    }
  }

  printf("answer1: %d\n", a1);
  printf("answer2: %d\n", a2);

  return 0;
}
